processing/src/main/java/org/apache/druid/query/groupby/strategy/GroupByStrategyV2.java

```diff
@@ -419,36 +436,81 @@ public Row apply(final Row row)
               false
           )
       );
+
+
       List<Sequence<Row>> subtotalsResults = new ArrayList<>(subtotals.size());
 
       Map<String, DimensionSpec> queryDimensionSpecs = new HashMap(queryWithoutSubtotalsSpec.getDimensions().size());
+      List<String> queryDimNames = new ArrayList(queryWithoutSubtotalsSpec.getDimensions().size());
+
       for (DimensionSpec dimSpec : queryWithoutSubtotalsSpec.getDimensions()) {
         queryDimensionSpecs.put(dimSpec.getOutputName(), dimSpec);
+        queryDimNames.add(dimSpec.getOutputName());
       }
 
       for (List<String> subtotalSpec : subtotals) {
         GroupByQuery subtotalQuery = queryWithoutSubtotalsSpec.withDimensionSpecs(
             subtotalSpec.stream()
-                        .map(s -> new DefaultDimensionSpec(s, s, queryDimensionSpecs.get(s).getOutputType()))
+                        .map(queryDimensionSpecs::get)
                         .collect(Collectors.toList())
         );
 
-        subtotalsResults.add(applyPostProcessing(
-            mergeResults(new QueryRunner<Row>()
-            {
-              @Override
-              public Sequence<Row> run(QueryPlus<Row> queryPlus, Map<String, Object> responseContext)
+        if (Utils.isPrefix(subtotalSpec, queryDimNames)) {
+          subtotalsResults.add(applyPostProcessing(
+              mergeResults(new QueryRunner<Row>()
               {
-                return GroupByRowProcessor.getRowsFromGrouper(
-                    queryWithoutSubtotalsSpec,
-                    subtotalSpec,
-                    grouperSupplier
-                );
-              }
-            }, subtotalQuery, null),
-            subtotalQuery
-                             )
-        );
+                @Override
+                public Sequence<Row> run(QueryPlus<Row> queryPlus, Map<String, Object> responseContext)
+                {
+                  return GroupByRowProcessor.getRowsFromGrouper(
+                      queryWithoutSubtotalsSpec,
+                      subtotalSpec,
+                      grouperSupplier
```

Comment:
This is `Grouper#1` right? Could you please add a comment where it's initialized, or change its name (`grouperOneSupplier` or something) to make the logic easier to follow?

Reply:
Or maybe breaking the method up would make it easier.

I guess what I'm trying to say is when reading, I was wishing it was more clear when we were using `Grouper#1`, when `Grouper#2`, and also who's responsible for closing what.

Reply:
renamed variable names to align them better with the comment, it should be easier to follow now.

I thought of breaking the method and then realized most of it is actually method calls which are big because of the number of arguments they take. take a look now, if it still looks difficult then I can try to break it.